### PR TITLE
Fix overlapping RangeDifferences during balanced text compare

### LINKED ISSUE #458: [Regression] Balance text compare causes overlapping RangeDifferences
#323 introduced a regression:

https://github.com/eclipse-platform/eclipse.platform/pull/323/files/8a8d01488ef4cea75efb981b806fdac1f3c04a07#diff-80809369e0a624c43bbae4fa6e3aca3bfc86cb3878320f4339bc0d9e2ae2aeb2R126

`previousEnd` is never incremented and `RangeDifference`s can now overlap.

This fails, e.g. when adding all such diffs to a `MultiTextEdit`:

```java
		MultiTextEdit result = new MultiTextEdit();

		TextRangeComparator left = new TextRangeComparator(fromText, compareMode);
		TextRangeComparator right = new TextRangeComparator(toText, compareMode);

		RangeDifference[] differences = RangeDifferencer.findDifferences(left, right);
		for (RangeDifference difference : differences) {
			TextRange leftRange = left.getRange(difference.leftStart(), difference.leftLength());
			TextRange rightRange = right.getRange(difference.rightStart(), difference.rightLength());

			ReplaceEdit replaceEdit = new ReplaceEdit(leftRange.offset, leftRange.length, rightRange.text);
			result.addChild(replaceEdit);
		}
		result.apply(document);

```

```
org.eclipse.text.edits.MalformedTreeException: Overlapping text edits
	at org.eclipse.text.edits.TextEdit$InsertionComparator.compare(TextEdit.java:126)
	at org.eclipse.text.edits.TextEdit$InsertionComparator.compare(TextEdit.java:1)
	at java.base/java.util.Collections.indexedBinarySearch(Collections.java:336)
	at java.base/java.util.Collections.binarySearch(Collections.java:324)
	at org.eclipse.text.edits.TextEdit.computeInsertionIndex(TextEdit.java:799)
	at org.eclipse.text.edits.TextEdit.internalAdd(TextEdit.java:784)
	at org.eclipse.text.edits.TextEdit.addChild(TextEdit.java:340)
	...
```

### PR DESCRIPTION
**Pull request by sratz**
Fixup for #66.
Fixes #458.

### CODE REVIEW

File: team/bundles/org.eclipse.compare.core/src/org/eclipse/compare/rangedifferencer/RangeDifferencer.java [hunk 1]
```diff
@@ -152,6 +152,7 @@ else if (diff.leftLength() == 0) {
 					optimized = factory.createRangeDifference(diff.kind(), diff.rightStart - moved, diff.rightLength, diff.leftStart - moved, diff.leftLength);
 				}
 			}
+			previousEnd = Math.max(optimized.rightEnd(), optimized.leftEnd());
```
Comment (sratz) [inline team/bundles/org.eclipse.compare.core/src/org/eclipse/compare/rangedifferencer/RangeDifferencer.java:155]: I am not 100% sure here, do we need two variables `previousLeftEnd` / `previousRightEnd` instead?

Comment (Bananeweizen) [reply team/bundles/org.eclipse.compare.core/src/org/eclipse/compare/rangedifferencer/RangeDifferencer.java:155]: It looks okay to me, but I have been thinking back and forth for 10 minutes now, too. When I wrote that code, it looked all so clear, and now it doesn't anymore...

Comment (sratz) [reply team/bundles/org.eclipse.compare.core/src/org/eclipse/compare/rangedifferencer/RangeDifferencer.java:155]: Hm, I can't get my head around it either.

I also have some more concerns, see #66

Comment (Bananeweizen) [reply team/bundles/org.eclipse.compare.core/src/org/eclipse/compare/rangedifferencer/RangeDifferencer.java:155]: I'm okay with reverting.

**Comment by sratz** [comment]
Closing in favor of revert: #461

**Pull request closed without merging**